sdk/include/ds/pointer.h

```diff
@@ -0,0 +1,256 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file Pointer utilities
+ */
+
+#pragma once
+
+#include <cheri.hh>
+#include <concepts>
+
+namespace ds::pointer
+{
+
+	/**
+	 * Offset a pointer by an unsigned number of bytes.  The return type must be
+	 * explicitly specified by the caller.
+	 */
+	template<typename T, typename U>
+	static inline __always_inline T *offset(U *base, size_t diff)
+	{
+		return reinterpret_cast<T *>(reinterpret_cast<uintptr_t>(base) + diff);
+	}
+
+	/**
+	 * Offset a pointer by a signed number of bytes.  The return type must be
+	 * explicitly specified by the caller.
+	 */
+	template<typename T, typename U>
+	static inline __always_inline T *offset_signed(U *base, ptrdiff_t diff)
```

Comment:
Is this not a different overload for the second parameter?  Does it need a different name?  Are there cases where it's ambiguous whether the `size_t` or `ptrdiff_t` version is intended at the call site (if so, please add a comment)?

Reply:
I was mostly cribbing from `snmalloc`, but I've switched to a single `offset()` that's templated on the type of the offset.